internal/databases/postgres/backup_push_handler.go

```diff
@@ -75,226 +141,437 @@ func getDeltaConfig() (maxDeltas int, fromFull bool) {
 	} else {
-		previousBackupName,
-			previousBackupSentinelDto,
-			incrementCount = configureDeltaBackup(uploader.UploadingFolder, deltaBaseSelector, isPermanent)
-	}
-
-	backupConfig := BackupConfig{
-		uploader:                  uploader,
-		archiveDirectory:          archiveDirectory,
-		backupsFolder:             utility.BaseBackupPath,
-		previousBackupName:        previousBackupName,
-		previousBackupSentinelDto: previousBackupSentinelDto,
-		isPermanent:               isPermanent,
-		forceIncremental:          false,
-		incrementCount:            incrementCount,
-		verifyPageChecksums:       verifyPageChecksums,
-		storeAllCorruptBlocks:     storeAllCorruptBlocks,
-		tarBallComposerType:       tarBallComposerType,
-		userData:                  userData,
-	}
-	createAndPushBackup(&backupConfig)
-}
-
-func configureDeltaBackup(folder storage.Folder, deltaBaseSelector internal.BackupSelector, isPermanent bool,
-) (string, BackupSentinelDto, int) {
+		err := bc.configureDeltaBackup()
+		tracelog.ErrorLogger.FatalOnError(err)
+	}
+
+	bc.createAndPushBackup()
+}
+
+func (bc *BackupHandler) createAndPushRemoteBackup() {
+	uploader := *bc.workers.uploader
+	uploader.UploadingFolder = uploader.UploadingFolder.GetSubFolder(utility.BaseBackupPath)
+	tracelog.DebugLogger.Printf("Uploading folder: %s", uploader.UploadingFolder)
+
+	baseBackup := bc.runRemoteBackup()
+	tracelog.InfoLogger.Println("Updating metadata")
+	bc.updateMetaData(uint64(baseBackup.StartLSN), baseBackup)
+	tracelog.InfoLogger.Println("Uploading metadata")
+	bc.uploadMetadata()
+	// logging backup set name
+	tracelog.InfoLogger.Printf("Wrote backup with name %s", bc.curBackupInfo.name)
+}
+
+func (bc *BackupHandler) uploadMetadata() {
+	curBackupName := bc.curBackupInfo.name
+	err := bc.uploadMetadataFile()
+	if err != nil {
+		tracelog.ErrorLogger.Printf("Failed to upload metadata file for backup: %s %v", curBackupName, err)
+		tracelog.ErrorLogger.FatalError(err)
+	}
+	err = bc.uploadSentinel()
+	if err != nil {
+		tracelog.ErrorLogger.Printf("Failed to upload sentinel file for backup: %s", curBackupName)
+		tracelog.ErrorLogger.FatalError(err)
+	}
+}
+
+// NewBackupHandler returns a backup handler object, which can handle the backup
+func NewBackupHandler(arguments BackupArguments) (bc *BackupHandler, err error) {
+	// RemoteBackup is triggered by not passing PGDATA to wal-g,
+	// and version cannot be read easily using replication connection.
+	// Retrieve both with this helper function which uses a temp connection to postgres.
+
+	uploader, err := ConfigureWalUploader()
+	if err != nil {
+		return bc, err
+	}
+	hostname, err := os.Hostname()
+	if err != nil {
+		return bc, err
+	}
+	pgInfo, err := getPgServerInfo()
+	if err != nil {
+		return bc, err
+	}
+
+	bc = &BackupHandler{
+		arguments: arguments,
+		workers: BackupWorkers{
+			uploader: uploader,
+		},
+		pgInfo: pgInfo,
+	}
+
+	bc.curBackupInfo.meta = ExtendedMetadataDto{
```

Comment:
This metadata is incomplete and it is easy to make a mistake when using and modifying it later. For example, in createAndPushBackup() we set some of these values for the second time.

I don't think it should be initialized here, because we don't have enough information yet. Let's strive to create the metadata files (ExtendedMetadataDto and BackupSentinelDto) after we have all the required information to build it to avoid such incomplete states.

Reply:
This started of with a lint error that there was too much going on in one func.
Which I then split into multiple funcs, and I needed to pass around a lot of vars.
By adding them into a struct managing the info got much easier and clearer.
And the I started changing funcs into methods and it all started to make sense.

The point is, I needed this as a placeholder for the data, which is sort of the other way around against initialise it only when you have all data.
Looking into the code there is two important moments: Start of backup (where a lot of info is registered) and end of end of backup (where the rest is registered before it is uploaded).
I am now looking into just having a func and a method to do these two parts.